overlord/snapstate/handlers.go

```diff
@@ -4814,3 +4814,162 @@ var getDirMigrationOpts = func(st *state.State, snapst *SnapState, snapsup *Snap
 
 	return opts, nil
 }
+
+func (m *SnapManager) doSetupKernelSnap(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	perfTimings := state.TimingsForTask(t)
+	snapsup, snapSt, err := snapSetupAndState(t)
+	if err != nil {
+		return err
+	}
+
+	st.Unlock()
+	pm := NewTaskProgressAdapterUnlocked(t)
+	timings.Run(perfTimings, "setup-kernel-snap",
+		fmt.Sprintf("setup of kernel snap %q", snapsup.InstanceName()),
+		func(timings.Measurer) {
+			err = m.backend.SetupKernelSnap(
+				snapsup.InstanceName(), snapsup.Revision(), pm)
+		})
+	st.Lock()
+	if err != nil {
+		return err
+	}
+
+	perfTimings.Save(st)
+
+	// Needed so the old drivers tree can be removed later
+	prevRev := snapSt.Current
+	t.Change().Set("previous-kernel-rev", prevRev)
+
+	// Make sure we won't be rerun
+	t.SetStatus(state.DoneStatus)
+
+	return nil
+}
+
+func (m *SnapManager) undoSetupKernelSnap(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	perfTimings := state.TimingsForTask(t)
+	snapsup, _, err := snapSetupAndState(t)
+	if err != nil {
+		return err
+	}
+
+	st.Unlock()
+	pm := NewTaskProgressAdapterUnlocked(t)
+	timings.Run(perfTimings, "cleanup-kernel-snap",
+		fmt.Sprintf("undo setup of kernel snap %q", snapsup.InstanceName()),
+		func(timings.Measurer) {
+			err = m.backend.CleanupKernelSnap(
+				snapsup.InstanceName(), snapsup.Revision(), pm)
+		})
+	st.Lock()
+	if err != nil {
+		return err
+	}
+
+	perfTimings.Save(st)
+	// Make sure we won't be rerun
+	t.SetStatus(state.UndoneStatus)
+
+	return nil
+}
+
+func (m *SnapManager) doCleanupOldKernelSnap(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	perfTimings := state.TimingsForTask(t)
+	_, snapst, err := snapSetupAndState(t)
+	if err != nil {
+		return err
+	}
+
+	currInfo, err := snapst.CurrentInfo()
+	if err != nil {
+		return err
+	}
+
+	// This is stored by doSetupKernelSnap - now after the reboot triggered
+	// after linking the new snap, we can remove the old drivers tree.
+	var prevRev snap.Revision
```

Comment:
do we need to actually double check that a reboot happened before proceeding with this?

Reply:
In theory this task should run only after the reboot (see my comment about restart boundaries), minus programming errors.

Reply:
If it's a concern we could do something similar to `snapstate.FinishRestart`, but with a proper name.